python/cuda_parallel/cuda/parallel/experimental/algorithms/scan.py

```diff
@@ -0,0 +1,165 @@
+# Copyright (c) 2024-2025, NVIDIA CORPORATION & AFFILIATES. ALL RIGHTS RESERVED.
+#
+#
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+from __future__ import annotations  # TODO: required for Python 3.7 docs env
+
+import ctypes
+from typing import Callable
+
+import numba
+import numpy as np
+from numba import cuda
+from numba.cuda.cudadrv import enums
+
+from .. import _cccl as cccl
+from .._bindings import get_bindings, get_paths
+from .._caching import CachableFunction, cache_with_key
+from .._utils import protocols
+from ..iterators._iterators import IteratorBase
+from ..typing import DeviceArrayLike, GpuStruct
+
+
+class _Scan:
+    # TODO: constructor shouldn't require concrete `d_in`, `d_out`:
+    def __init__(
+        self,
+        d_in: DeviceArrayLike | IteratorBase,
+        d_out: DeviceArrayLike | IteratorBase,
+        op: Callable,
+        h_init: np.ndarray | GpuStruct,
+    ):
+        # Referenced from __del__:
+        self.build_result = None
+
+        self.d_in_cccl = cccl.to_cccl_iter(d_in)
+        self.d_out_cccl = cccl.to_cccl_iter(d_out)
+        self.h_init_cccl = cccl.to_cccl_value(h_init)
+        cc_major, cc_minor = cuda.get_current_device().compute_capability
+        cub_path, thrust_path, libcudacxx_path, cuda_include_path = get_paths()
+        if isinstance(h_init, np.ndarray):
+            value_type = numba.from_dtype(h_init.dtype)
+        else:
+            value_type = numba.typeof(h_init)
+        sig = (value_type, value_type)
+        self.op_wrapper = cccl.to_cccl_op(op, sig)
+        self.build_result = cccl.DeviceScanBuildResult()
+        self.bindings = get_bindings()
+        error = self.bindings.cccl_device_scan_build(
+            ctypes.byref(self.build_result),
+            self.d_in_cccl,
+            self.d_out_cccl,
+            self.op_wrapper,
+            cccl.to_cccl_value(h_init),
+            cc_major,
+            cc_minor,
+            ctypes.c_char_p(cub_path),
+            ctypes.c_char_p(thrust_path),
+            ctypes.c_char_p(libcudacxx_path),
+            ctypes.c_char_p(cuda_include_path),
+        )
+        if error != enums.CUDA_SUCCESS:
+            raise ValueError("Error building scan")
+
+    def __call__(
+        self,
+        temp_storage,
+        d_in,
+        d_out,
+        num_items: int,
+        h_init: np.ndarray | GpuStruct,
+        stream=None,
+    ):
+        if self.d_in_cccl.type.value == cccl.IteratorKind.POINTER:
+            self.d_in_cccl.state = protocols.get_data_pointer(d_in)
+        else:
+            self.d_in_cccl.state = d_in.state
+
+        if self.d_out_cccl.type.value == cccl.IteratorKind.POINTER:
+            self.d_out_cccl.state = protocols.get_data_pointer(d_out)
+        else:
+            self.d_out_cccl.state = d_out.state
+
+        self.h_init_cccl.state = h_init.__array_interface__["data"][0]
```

Comment:
Will this line work for `h_init` of type `GpuStruct`?

Reply:
Yes, because of https://github.com/NVIDIA/cccl/blob/62b7a8e883e6528d8d3803f73f775571e60edc3a/python/cuda_parallel/cuda/parallel/experimental/struct.py#L74-L75, which is a hack to be sure, but it works for now.

Reply:
I see: 

```
In [1]: from cuda.parallel.experimental.struct import gpu_struct

In [2]: import numpy as np

In [3]: @gpu_struct
   ...: class Pointer:
   ...:     x : np.float32
   ...:     y : np.float32
   ...: 

In [4]: p = Pointer(3, 4)

In [5]: p.__array_interface__
Out[5]: 
{'data': (109703859879504, False),
 'strides': None,
 'descr': [('x', '<f4'), ('y', '<f4')],
 'typestr': '|V8',
 'shape': (1,),
 'version': 3}
```